app/services/authentication/providers.rb

```diff
@@ -1,11 +1,70 @@
+# We require all authentication modules to make sure providers
+# are correctly preloaded both in development and in production and
+# ready to be used when needed at runtime
```

Comment:
> to make sure providers are correctly preloaded both in development and in production and ready to be used when needed at runtime 

Shouldn't Rails do this for us? I read up a bit on `require_dependency` and it seems like it should rarely be used. I guess, even with this comment, I am still confused why we explicitly need it here. Maybe I'm just missing something?

Reply:
it doesn't auto load them unfotunately as I need to know which subclasses `Provider` has to load them as plugins. This is the previous version of that method:

https://github.com/thepracticaldev/dev.to/pull/7505/commits/9f0d2c08956630e0e4dbf64840707d0710ec603c#diff-24cf9abb14a9d705c426c66026c693eaR27-R35

and this is the conversation I had with Michael about it https://github.com/thepracticaldev/dev.to/pull/7505#discussion_r415491037

Reply:
Like I said, in production, it should always work. In development you'll experience what you described, due to eager loading. `require_dependency` is generally the way to resolve this while still being part of Rails' usual constant tracking for reloading etc.